code/messaging/src/main/java/com/adobe/marketing/mobile/messaging/PresentableMessageMapper.java

```diff
@@ -0,0 +1,295 @@
+/*
+  Copyright 2024 Adobe. All rights reserved.
+  This file is licensed to you under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License. You may obtain a copy
+  of the License at http://www.apache.org/licenses/LICENSE-2.0
+  Unless required by applicable law or agreed to in writing, software distributed under
+  the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+  OF ANY KIND, either express or implied. See the License for the specific language
+  governing permissions and limitations under the License.
+ */
+package com.adobe.marketing.mobile.messaging;
+
+import static com.adobe.marketing.mobile.messaging.MessagingConstants.EventDataKeys.RulesEngine.MESSAGE_CONSEQUENCE_DETAIL_KEY_CONTENT;
+import static com.adobe.marketing.mobile.messaging.MessagingConstants.EventDataKeys.RulesEngine.MESSAGE_CONSEQUENCE_DETAIL_KEY_DATA;
+import static com.adobe.marketing.mobile.messaging.MessagingConstants.EventDataKeys.RulesEngine.MESSAGE_CONSEQUENCE_DETAIL_KEY_SCHEMA;
+import static com.adobe.marketing.mobile.messaging.MessagingConstants.SchemaValues.SCHEMA_IAM;
+
+import androidx.annotation.Nullable;
+
+import com.adobe.marketing.mobile.Message;
+import com.adobe.marketing.mobile.MessagingEdgeEventType;
+import com.adobe.marketing.mobile.launch.rulesengine.RuleConsequence;
+import com.adobe.marketing.mobile.services.Log;
+import com.adobe.marketing.mobile.services.ServiceProvider;
+import com.adobe.marketing.mobile.services.ui.InAppMessage;
+import com.adobe.marketing.mobile.services.ui.Presentable;
+import com.adobe.marketing.mobile.services.ui.UIService;
+import com.adobe.marketing.mobile.services.ui.message.InAppMessageSettings;
+import com.adobe.marketing.mobile.util.DataReader;
+import com.adobe.marketing.mobile.util.DefaultPresentationUtilityProvider;
+import com.adobe.marketing.mobile.util.MapUtils;
+import com.adobe.marketing.mobile.util.StringUtils;
+
+import java.lang.ref.WeakReference;
+import java.util.HashMap;
+import java.util.Map;
+
+public final class PresentableMessageMapper {
```

Comment:
Discussed with Spoorthi on a call about modifying this class to prevent exposing the mapping logic. 

The reason a mapping is being held between the Presentable `id` and the `Message` is because customers who implement a `PresentationDelegate` can chose to manage the display and dismissal in `PresentationDelegate.canShow()` and also use it to enable and disable tracking for the Message associated with the Presentable.  The app implementer does not necessarily need to know about the mapping.
Instead of exposing the `PresentableMessageMapper` and the underlying mapping employed, we agreed that a simple utility method that provides a `Message` given a `Presentable<InAppMessage>` is sufficient for this use case.
 
```
class MessageUtilities { // or some appropriate name
     ....
     fun getMessageFromPresentable(@NonNull Presentable<InAppMessage> presentable): Message? {
        val id = getPresentation().getId()
        return PresentableMapper.getMessageFromPresentableId(id)
     }
}
```

There is a need to ensure that the map is cleaned up, but since the `PresentationDelegate` implementer can hold references to the `Presentable<InAppMessage>` and can invoke `show()`/`dismiss()` later (and likely need the id mapping) we cannot cleanup on show or on error. Given that the accumulation of several presentables is not common (perhaps rare also), we plan to check for the references that went out of context while creating new InAppMessages in the extension.

This allows us to make PresentableMessageMapper private to the extension and limit exposing the internal logic.

Reply:
Changes as per discussion